module/zfs/zfs_vnops.c

```diff
@@ -3656,8 +3656,7 @@ int
 zfs_rename(struct inode *sdip, char *snm, struct inode *tdip, char *tnm,
```

Comment:
Rather than modify this existing complicated function so extensively, what if instead you added a new `zfs_renameat2` which is solely responsible for handling `RENAME_EXCHANGE` and `TX_WHITEOUT`.  That would potentially allow it to be simpler and remove the risk of accidentally breaking `zfs_rename` in some subtle way.  (Or even a `zfs_renameat2_exchange()` and `zfs_renameat2_whiteout` if that falls out more cleanly).

Reply:
I could give this a shot, though there is a fair amount of shared code -- the main difference is right at the end in `zfs_rename`.